store/primary/primary.go

```diff
@@ -109,55 +173,53 @@ func (s *rtStorage) put(k string, entry store.IndexEntry) bool {
 		return false
 	}
 
-	s.current.Put(k, append(old, entry))
+	c.current.Put(k, append(old, entry))
 	return true
 }
 
-func (s *rtStorage) rotateCache() {
-	// Acquire all locks
-	for i := range s.branchLocks {
-		s.branchLocks[i].Lock()
-	}
-
-	if s.current.Len() >= s.sizeLimit {
-		s.previous, s.current = s.current, radixtree.New()
-	}
-
-	// Release all locks
-	for i := range s.branchLocks {
-		s.branchLocks[i].Unlock()
+func (c *syncCache) remove(k string, entry store.IndexEntry) bool {
+	removed := removeEntry(c.current, k, entry)
+	if removeEntry(c.previous, k, entry) {
+		removed = true
 	}
+	return removed
 }
 
-// PutMany puts store.IndexEntry information in several CIDs.
-// This is usually triggered when a bulk update for a providerID-pieceID
-// arrives.
-func (s *rtStorage) PutMany(cids []cid.Cid, provID peer.ID, pieceID cid.Cid) error {
-	in := store.IndexEntry{ProvID: provID, PieceID: pieceID}
-	var addedEntry bool
-
-	for i := range cids {
-		k := string(cids[i].Hash())
-		s.lockBranch(k)
+func removeEntry(cache *radixtree.Bytes, k string, entry store.IndexEntry) bool {
+	// Get from current cache
+	v, found := cache.Get(k)
+	if !found {
+		return false
+	}
 
-		if s.put(k, in) {
-			addedEntry = true
+	values := v.([]store.IndexEntry)
+	for i := range values {
+		if entry.PieceID == values[i].PieceID &&
+			entry.ProvID == values[i].ProvID {
+			if len(values) == 1 {
+				cache.Delete(k)
+			} else {
+				values[i] = values[len(values)-1]
+				values[len(values)-1] = store.IndexEntry{}
+				cache.Put(k, values[:len(values)-1])
+			}
+			return true
 		}
-		s.unlockBranch(k)
 	}
+	return false
+}
 
-	if addedEntry {
-		// Check if rotation needed
-		s.lockBranch("")
-		if s.current.Len() >= s.sizeLimit {
-			s.unlockBranch("")
-			s.rotateCache()
-		} else {
-			s.unlockBranch("")
-		}
+func (c *syncCache) removeAll(k string) bool {
```

Comment:
Is this function being used somewhere? I don't see it exposed through the interface nor used in any other function.

Reply:
No currently.  I was anticipating its use, but I will remove it since I do not think it will get used since I do not know under what conditions we will want to remove a CID without knowing which provider provides it.